### PR TITLE
Add test for sycl::has_kernel_bundle for "core" test case

### PR DESCRIPTION
**Pull request by vasilytric**
These tests verifies that result of `sycl::has_kernel_bundle` is equal to compatible status for pair of current kernel requirements and tested device

There are so many files due to part of these tests should be launched in separate `*.exe`-file and another tests splitted into different files for have possible to know which test are failed
if we have all tests into one file then if just one test failed all tests will be marked as failed and pass rate will be decreased (but this pass rate is not correspond to the real pass rate)

___________________________________________
This PR depend on these PRs:
- https://github.com/KhronosGroup/SYCL-CTS/pull/218
- https://github.com/KhronosGroup/SYCL-CTS/pull/220

The `SYCL CTS CI / compile-cts (dpcpp)` failed due to `add_independent_cts_tests` CMake function is not defined in this PR, it can be found [here](https://github.com/KhronosGroup/SYCL-CTS/pull/218/files#diff-4461c617ceae0c7e0206622aacdf555aedd62eb129c2efb74c84fa1567bcbe0dR197-R214)

### CODE REVIEW

**Review by ProGTX — APPROVED** [review]

File: tests/kernel_bundle/has_kernel_bundle_core.h [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+ */
+template <typename KernelDescriptorT, sycl::bundle_state BundleState>
+struct core<KernelDescriptorT, BundleState, overload::id::ctx_only> {
+  void operator()(sycl_cts::util::logger &log, const sycl::context &ctx,
+                  const sycl::device &dev, const std::string &kernel_name) {
+    using kernel_functor = typename KernelDescriptorT::type;
+    auto k_id{sycl::get_kernel_id<kernel_functor>()};
+    auto restrictions{
+        kernel_bundle::get_restrictions<KernelDescriptorT, BundleState>()};
+    std::string compat_info{};
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_core.h]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/223/commits/ffe8c589982479f4f8189d35896c36b83b3c97bd#diff-598f30f03fe7c5cb0502b789f2fc4402424e4f760d531254fc3f44b30ddacf80L130-L131)

File: tests/kernel_bundle/has_kernel_bundle_core.h [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+                ", but device compatible status: " +
+                sycl_cts::get_cts_string::for_bool(dev_is_compat) + ". for " +
+                kernel_name + ", with restrictions:" +
+                string_with_restrictions + " and device " + compatible_info);
+}
+
+/** @brief Verify has_kernel_bundle result by rules that depend on
+ *         sycl::bundle_state
+ *  @details If we got unexpected has_kernel_bundle result then this function
+ *           return TRUE
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_core.h]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/223/commits/ffe8c589982479f4f8189d35896c36b83b3c97bd#diff-598f30f03fe7c5cb0502b789f2fc4402424e4f760d531254fc3f44b30ddacf80L88-L89)

File: tests/kernel_bundle/has_kernel_bundle_core.h [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+ *  @tparam BundleStateT sycl::bundle_state enum's enumeration's field
+ *  @param has_kb_result Result of sycl::has_kernel_bundle calling
+ *  @param dev_is_compat Current device compatible status
+ *  @retval Bool value that correspond to has_kernel_bundle result that NOT
+ *          equal to device compatible status according to rules that depend on
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_core.h]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/223/commits/ffe8c589982479f4f8189d35896c36b83b3c97bd#diff-598f30f03fe7c5cb0502b789f2fc4402424e4f760d531254fc3f44b30ddacf80L92-R95)

File: tests/kernel_bundle/has_kernel_bundle_core.h [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+ */
+template <typename KernelDescriptorT, sycl::bundle_state BundleState>
+struct core<KernelDescriptorT, BundleState, overload::id::ctx_kid> {
+  void operator()(sycl_cts::util::logger &log, const sycl::context &ctx,
+                  const sycl::device &dev, const std::string &kernel_name) {
+    using kernel_functor = typename KernelDescriptorT::type;
+    auto k_id{sycl::get_kernel_id<kernel_functor>()};
+    auto restrictions{
+        kernel_bundle::get_restrictions<KernelDescriptorT, BundleState>()};
+    std::string compat_info{};
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_core.h]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/223/commits/ffe8c589982479f4f8189d35896c36b83b3c97bd#diff-598f30f03fe7c5cb0502b789f2fc4402424e4f760d531254fc3f44b30ddacf80L162-L163)

File: tests/kernel_bundle/has_kernel_bundle_core.h [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+ */
+template <typename KernelDescriptorT, sycl::bundle_state BundleState>
+struct core<KernelDescriptorT, BundleState, overload::id::ctx_dev> {
+  void operator()(sycl_cts::util::logger &log, const sycl::context &ctx,
+                  const sycl::device &dev, const std::string &kernel_name) {
+    using kernel_functor = typename KernelDescriptorT::type;
+    auto k_id{sycl::get_kernel_id<kernel_functor>()};
+    auto restrictions{
+        kernel_bundle::get_restrictions<KernelDescriptorT, BundleState>()};
+    std::string compat_info{};
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_core.h]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/223/commits/ffe8c589982479f4f8189d35896c36b83b3c97bd#diff-598f30f03fe7c5cb0502b789f2fc4402424e4f760d531254fc3f44b30ddacf80L194-L195)

File: tests/kernel_bundle/has_kernel_bundle_core.h [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+ */
+template <typename KernelDescriptorT, sycl::bundle_state BundleState>
+struct core<KernelDescriptorT, BundleState, overload::id::ctx_dev_kname> {
+  void operator()(sycl_cts::util::logger &log, const sycl::context &ctx,
+                  const sycl::device &dev, const std::string &kernel_name) {
+    using kernel_functor = typename KernelDescriptorT::type;
+    auto k_id = sycl::get_kernel_id<kernel_functor>();
+    auto restrictions{
+        kernel_bundle::get_restrictions<KernelDescriptorT, BundleState>()};
+    std::string compat_info{};
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_core.h]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/223/commits/ffe8c589982479f4f8189d35896c36b83b3c97bd#diff-598f30f03fe7c5cb0502b789f2fc4402424e4f760d531254fc3f44b30ddacf80L292-L293)

File: tests/kernel_bundle/has_kernel_bundle_core.h [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+ */
+template <typename KernelDescriptorT, sycl::bundle_state BundleState>
+struct core<KernelDescriptorT, BundleState, overload::id::ctx_kname> {
+  void operator()(sycl_cts::util::logger &log, const sycl::context &ctx,
+                  const sycl::device &dev, const std::string &kernel_name) {
+    using kernel_functor = typename KernelDescriptorT::type;
+    auto k_id{sycl::get_kernel_id<kernel_functor>()};
+    auto restrictions{
+        kernel_bundle::get_restrictions<KernelDescriptorT, BundleState>()};
+    std::string compat_info{};
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_core.h]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/223/commits/ffe8c589982479f4f8189d35896c36b83b3c97bd#diff-598f30f03fe7c5cb0502b789f2fc4402424e4f760d531254fc3f44b30ddacf80L259-L260)

File: tests/kernel_bundle/has_kernel_bundle_core.h [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+ */
+template <typename KernelDescriptorT, sycl::bundle_state BundleState>
+struct core<KernelDescriptorT, BundleState, overload::id::ctx_dev_kid> {
+  void operator()(sycl_cts::util::logger &log, const sycl::context &ctx,
+                  const sycl::device &dev, const std::string &kernel_name) {
+    using kernel_functor = typename KernelDescriptorT::type;
+    auto k_id{sycl::get_kernel_id<kernel_functor>()};
+    auto restrictions{
+        kernel_bundle::get_restrictions<KernelDescriptorT, BundleState>()};
+    std::string compat_info{};
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_core.h]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/223/commits/ffe8c589982479f4f8189d35896c36b83b3c97bd#diff-598f30f03fe7c5cb0502b789f2fc4402424e4f760d531254fc3f44b30ddacf80L226-L227)

**Review by keryell — APPROVED** [review]
Thanks!

**Comment by bader** [comment]
@vasilytric, please, resolve merge conflicts.

Also, there is a compilation failure, which looks like a bug in the test.

>  ../tests/kernel_bundle/has_kernel_bundle_core.h:77:50: error: invalid operands to binary expression ('const char[35]' and 'std::string_view' (aka 'basic_string_view<char>'))

**Comment by bader** [comment]
@vasilytric, please, resolve merge conflicts.

**Pull request merged**